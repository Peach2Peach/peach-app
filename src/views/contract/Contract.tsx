import { Header, PeachScrollView, Screen } from '../../components'
import { TradeSummary } from '../../components/offer'
import tw from '../../styles/tailwind'

import { useMemo } from 'react'
import { useRoute, useShowHelp } from '../../hooks'
import { useConfirmCancelTrade } from '../../popups/tradeCancelation'
import { canCancelContract, contractIdToHex } from '../../utils/contract'
import { headerIcons } from '../../utils/layout'
import { LoadingScreen } from '../loading/LoadingScreen'
import { ContractActions } from './ContractActions'
import { ContractContext, useContractContext } from './context'
import { useContractSetup } from './hooks/useContractSetup'

export const Contract = () => {
  const { contract, isLoading, view, showBatchInfo, toggleShowBatchInfo, ...contractActionsProps } = useContractSetup()

  if (!contract || !view || isLoading) return <LoadingScreen />

  return (
    <ContractContext.Provider value={{ contract, view, showBatchInfo, toggleShowBatchInfo }}>
      <Screen header={<ContractHeader requiredAction={contractActionsProps.requiredAction} />}>
        <PeachScrollView contentContainerStyle={tw`grow`} contentStyle={tw`grow`}>
          <TradeSummary />
          <ContractActions style={tw`items-center justify-end w-full`} {...contractActionsProps} />
        </PeachScrollView>
      </Screen>
    </ContractContext.Provider>
  )
}

type Props = {
  requiredAction: ContractAction
}

const themes = {
  buyer: tw`bg-success-background-dark border-success-mild-1 text-success-main`,
  seller: tw`bg-primary-background-dark border-primary-mild-1 text-primary-main`,
  paymentTooLate: tw`bg-warning-mild-1 border-warning-mild-2 text-black-1`,
  dispute: tw`bg-error-main border-error-dark text-primary-background-light`,
  cancel: tw`bg-black-5 border-black-4 text-black-2`,
}

function ContractHeader ({ requiredAction }: Props) {
  const { contractId } = useRoute<'contract'>().params
  const { contract, view } = useContractContext()
  const { showConfirmPopup } = useConfirmCancelTrade()
  const showMakePaymentHelp = useShowHelp('makePayment')
  const showConfirmPaymentHelp = useShowHelp('confirmPayment')

  const memoizedIcons = useMemo(() => {
    const icons = []
    if (contract && canCancelContract(contract, view)) icons.push({
      ...headerIcons.cancel,
      onPress: () => showConfirmPopup(contract),
    })
    if (view === 'buyer' && requiredAction === 'sendPayment') icons.push({
      ...headerIcons.help,
      onPress: showMakePaymentHelp,
    })
    if (view === 'seller' && requiredAction === 'confirmPayment') icons.push({
      ...headerIcons.help,
      onPress: showConfirmPaymentHelp,
    })
    return contract?.disputeActive ? [] : icons
  }, [showConfirmPopup, contract, requiredAction, showConfirmPaymentHelp, showMakePaymentHelp, view])

  return (
    <Header
      icons={memoizedIcons}
      title={contractIdToHex(contractId)}
      style={[tw`border-b rounded-b-lg`, themes.dispute]}
      theme="buyer"
      subtitle={<Header.Subtitle amount={contract.amount} premium={contract.premium} viewer={view} theme="buyer" />}
    />
  )
}
