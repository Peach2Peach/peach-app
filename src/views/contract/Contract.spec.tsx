import Contract from './Contract'
import { createRenderer } from 'react-test-renderer/shallow'

const defaultReturnValue = {
  contract: {
    buyer: {
      username: 'buyerUsername',
    },
    seller: {
      username: 'sellerUsername',
    },
    disputeActive: false,
  },
  isEmailRequired: false,
  isLoading: false,
  view: 'buyer',
  requiredAction: 'requiredAction',
  actionPending: false,
  postConfirmPaymentBuyer: jest.fn(),
  postConfirmPaymentSeller: jest.fn(),
}

const useContractSetupMock = jest.fn(() => defaultReturnValue)

jest.mock('./hooks/useContractSetup', () => ({
  useContractSetup: () => useContractSetupMock(),
}))

describe('Contract', () => {
  const renderer = createRenderer()
  it('should render a default Contract', () => {
    renderer.render(<Contract />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should show the loading screen if isLoading is true', () => {
    useContractSetupMock.mockReturnValueOnce({
      ...defaultReturnValue,
      isLoading: true,
    })
    renderer.render(<Contract />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should show the buyers profile overview if the viewer is the seller', () => {
    useContractSetupMock.mockReturnValueOnce({
      ...defaultReturnValue,
      view: 'seller',
    })
    renderer.render(<Contract />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should show the provide email button when the contract requires an email', () => {
    useContractSetupMock.mockReturnValueOnce({
      ...defaultReturnValue,
      contract: {
        ...defaultReturnValue.contract,
        // @ts-ignore
        isEmailRequired: true,
      },
    })
    renderer.render(<Contract />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should show the dispute sliders when the contract is in the refundOrReviveRequired state and there is a dispute winner', () => {
    useContractSetupMock.mockReturnValueOnce({
      ...defaultReturnValue,
      contract: {
        ...defaultReturnValue.contract,
        tradeStatus: 'refundOrReviveRequired',
        disputeWinner: 'seller',
      },
    })
    renderer.render(<Contract />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should not show the dispute sliders when there is no dispute winner', () => {
    useContractSetupMock.mockReturnValueOnce({
      ...defaultReturnValue,
      contract: {
        ...defaultReturnValue.contract,
        tradeStatus: 'refundOrReviveRequired',
        disputeWinner: undefined,
      },
    })
    renderer.render(<Contract />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
