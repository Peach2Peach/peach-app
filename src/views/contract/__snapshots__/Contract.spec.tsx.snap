// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract should render a default Contract 1`] = `
<Context.Provider
  value={
    {
      "contract": {
        "buyer": {
          "username": "buyerUsername",
        },
        "disputeActive": false,
        "seller": {
          "username": "sellerUsername",
        },
      },
      "showBatchInfo": undefined,
      "toggleShowBatchInfo": undefined,
      "view": "buyer",
    }
  }
>
  <Screen
    header={
      <ContractHeader
        requiredAction="requiredAction"
      />
    }
  >
    <ForwardRef
      contentContainerStyle={
        {
          "flexGrow": 1,
        }
      }
      contentStyle={
        {
          "flexGrow": 1,
        }
      }
    >
      <TradeSummary />
      <ContractActions
        actionPending={false}
        isEmailRequired={false}
        postConfirmPaymentBuyer={[MockFunction]}
        postConfirmPaymentSeller={[MockFunction]}
        requiredAction="requiredAction"
        style={
          {
            "alignItems": "center",
            "justifyContent": "flex-end",
            "width": "100%",
          }
        }
      />
    </ForwardRef>
  </Screen>
</Context.Provider>
`;

exports[`Contract should show the loading screen if isLoading is true 1`] = `<LoadingScreen />`;
